pkg/mappings.bzl

```diff
@@ -426,3 +428,166 @@ pkg_mkdirs = rule(
     },
     provides = [PackageDirsInfo],
 )
+
+def _pkg_mklink_impl(ctx):
+    out_attributes = json.decode(ctx.attr.attributes)
+
+    # The least surprising default mode is that of a symbolic link (0777).
+    # Permissions on symlinks typically don't matter, as the operation is
+    # typically moved to where the link is pointing.
+    out_attributes.setdefault("mode", "0777")
+    return [
+        PackageSymlinkInfo(
+            destination = ctx.attr.dest,
+            source = ctx.attr.src,
+            attributes = out_attributes,
+        ),
+    ]
+
+pkg_mklink = rule(
+    doc = """Define a symlink  within packages
+
+    This rule results in the creation of a single link within a package.
+
+    Symbolic links specified by this rule may point at files/directories outside of the
+    package, or otherwise left dangling.
+
+    """,
+    implementation = _pkg_mklink_impl,
+    # @unsorted-dict-items
+    attrs = {
+        "dest": attr.string(
+            doc = """Link "target", a path within the package.
+
+            This is the actual created symbolic link.
+
+            If the directory structure provided by this attribute is not
+            otherwise created when exist within the package when it is built, it
+            will be created implicitly, much like with `pkg_files`.
+
+            This path may be prefixed or rooted by grouping or packaging rules.
+
+            """,
+            mandatory = True,
+        ),
+        "src": attr.string(
+            doc = """Link "source", a path on the filesystem.
+
+            This is what the link "points" to, and may point to an arbitrary
+            filesystem path, even relative paths.
+
+            """,
+            mandatory = True,
+        ),
+        "attributes": attr.string(
+            doc = """Attributes to set on packaged symbolic links.
+
+            Always use `pkg_attributes()` to set this rule attribute.
+
+            Symlink permissions may have different meanings depending on your
+            host operating system; consult its documentation for more details.
+
+            If not otherwise overridden, the link's mode will be set to UNIX
+            "0777", or the target platform's equivalent.
+
+            Consult the "Mapping Attributes" documentation in the rules_pkg
+            reference for more details.
+            """,
+            default = "{}",  # Empty JSON
+        ),
+    },
+    provides = [PackageSymlinkInfo],
+)
+
+def _pkg_filegroup_impl(ctx):
+    files = []
+    dirs = []
+    links = []
+    mapped_files_depsets = []
+
+    if ctx.attr.prefix:
+        # If "prefix" is provided, we need to manipulate the incoming providers.
+        for s in ctx.attr.srcs:
+            if PackageFilesInfo in s:
+                new_pfi = PackageFilesInfo(
+                    dest_src_map = {
+                        paths.join(ctx.attr.prefix, dest): src
```

Comment:
Is paths OS specific or does it just use / as a sep?   I think we need to use / so that we can do the OS specific convert at package write time.

Reply:
`paths` always uses "/".  Quoth the [source](https://github.com/bazelbuild/bazel-skylib/blob/f80bc733d4b9f83d427ce3442be2e07427b2cc8d/lib/paths.bzl):

> NOTE: The functions in this module currently only support paths with Unix-style
path separators (forward slash, "/"); they do not handle Windows-style paths
with backslash separators or drive letters.

You mentioned in the call the morning regarding whether we needed `paths` from bazel-skylib.  I think we do -- at the very least we use `normalize`, `relativize`, and `join`.

We might be able to get away with vendoring it.  I'm not sure that's a good idea, but it can be considered later.